File: algorithms/graph/iterative-dfs/DFS.go
package main

import "fmt"

func iterDfs(graph map[string][]string, start string) {
	visited := make(map[string]bool)

	// Using slice as stack to simulate behavior
	stack := []string{start}

	for len(stack) > 0 {
		n := len(stack) - 1
		node := stack[n]
		stack = stack[:n]

		if visited[node] {
			continue
		}

		// For unvisited node ... finding neighbours and pushing them
		// into slice in reverse order to retain FIFO behavior when grabbing
		// the rear element in subsequent iteration
		visited[node] = true
		fmt.Println(node)
		neighbours := graph[node]

		for i := len(neighbours) - 1; i >= 0; i-- {
			neighbour := neighbours[i]

			if !visited[neighbour] {
				stack = append(stack, neighbour)
			}
		}
	}
}

func main() {
	graph := map[string][]string{
		"A": {"B", "C"},
		"B": {"D", "E"},
		"C": {"F"},
		"D": {},
		"E": {"F"},
		"F": {},
	}

	iterDfs(graph, "A")
}

File: algorithms/graph/iterative-dfs/DFS.cpp
#include <stack>
#include <vector>
using namespace std;

vector<int> depthFirstSearch(const vector<vector<int>> &graph, int start) {
  vector<int> traversal;
  vector<bool> visited(graph.size(), false);
  stack<int> nodes;

  nodes.push(start);

  while (!nodes.empty()) {
    int node = nodes.top();
    nodes.pop();

    if (visited[node]) {
      continue;
    }

    visited[node] = true;
    traversal.push_back(node);

    for (auto neighbor = graph[node].rbegin(); neighbor != graph[node].rend();
         ++neighbor) {
      if (!visited[*neighbor]) {
        nodes.push(*neighbor);
      }
    }
  }

  return traversal;
}
